File: src/risk_matrix.js
/**
 * Risk matrix logic: scoring, color mapping, and state management.
 */

export const RISK_COLORS = {
  critical: '#dc2626',
  high:     '#f97316',
  medium:   '#eab308',
  low:      '#22c55e',
};

export function getRiskLevel(score) {
  if (score >= 20) return 'critical';
  if (score >= 12) return 'high';
  if (score >= 6)  return 'medium';
  return 'low';
}

export function getRiskColor(score) {
  return RISK_COLORS[getRiskLevel(score)];
}

export function calcScore(likelihood, impact) {
  return likelihood * impact;
}

/**
 * Generate a 5x5 matrix cell descriptor array.
 * Returns: { likelihood, impact, score, level, color }
 */
export function buildMatrixCells() {
  const cells = [];
  for (let impact = 5; impact >= 1; impact--) {
    for (let likelihood = 1; likelihood <= 5; likelihood++) {
      const score = calcScore(likelihood, impact);
      cells.push({
        likelihood,
        impact,
        score,
        level: getRiskLevel(score),
        color: getRiskColor(score),
      });
    }
  }
  return cells;
}

/**
 * In-memory risk register.
 */
export class RiskRegister {
  constructor() {
    this._risks = [];
    this._nextId = 1;
  }

  add(risk) {
    const entry = {
      id: `R${String(this._nextId++).padStart(3, '0')}`,
      ...risk,
      score: calcScore(risk.likelihood, risk.impact),
      level: getRiskLevel(calcScore(risk.likelihood, risk.impact)),
      createdAt: new Date().toISOString(),
    };
    this._risks.push(entry);
    return entry;
  }

  remove(id) {
    this._risks = this._risks.filter((r) => r.id !== id);
  }

  update(id, updates) {
    const idx = this._risks.findIndex((r) => r.id === id);
    if (idx === -1) return;
    const risk = { ...this._risks[idx], ...updates };
    risk.score = calcScore(risk.likelihood, risk.impact);
    risk.level = getRiskLevel(risk.score);
    this._risks[idx] = risk;
  }

  getAll() {
    return [...this._risks].sort((a, b) => b.score - a.score);
  }

  getByCell(likelihood, impact) {
    return this._risks.filter(
      (r) => r.likelihood === likelihood && r.impact === impact
    );
  }

  getSummary() {
    const all = this._risks;
    return {
      total: all.length,
      critical: all.filter((r) => r.level === 'critical').length,
      high:     all.filter((r) => r.level === 'high').length,
      medium:   all.filter((r) => r.level === 'medium').length,
      low:      all.filter((r) => r.level === 'low').length,
    };
  }

  loadFromJSON(data) {
    for (const r of data.risks ?? []) {
      this.add(r);
    }
  }

  toJSON() {
    return {
      exportedAt: new Date().toISOString(),
      risks: this.getAll(),
    };
  }

  toCSV() {
    const header = 'ID,Title,Category,Likelihood,Impact,Score,Level,Owner,Status,Due Date';
    const rows = this.getAll().map((r) =>
      [
        r.id, `"${(r.title ?? '').replace(/"/g, '""')}"`,
        r.category ?? '',
        r.likelihood, r.impact, r.score,
        r.level,
        `"${(r.owner ?? '').replace(/"/g, '""')}"`,
        r.status ?? '',
        r.due_date ?? '',
      ].join(',')
    );
    return [header, ...rows].join('\n');
  }
}

File: python/risk_matrix.py
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Literal, Optional

RiskLevel = Literal["critical", "high", "medium", "low"]


def get_level(score: int) -> RiskLevel:
    if score >= 20: return "critical"
    if score >= 12: return "high"
    if score >= 6:  return "medium"
    return "low"


def get_color(level: RiskLevel) -> str:
    return {"critical": "#dc2626", "high": "#f97316", "medium": "#eab308", "low": "#22c55e"}[level]


@dataclass
class Risk:
    id: str
    title: str
    category: str
    likelihood: int
    impact: int
    description: str = ""
    existing_controls: str = ""
    mitigation: str = ""
    owner: str = ""
    due_date: str = ""
    status: str = "open"

    @property
    def score(self) -> int:
        return self.likelihood * self.impact

    @property
    def level(self) -> RiskLevel:
        return get_level(self.score)

    @property
    def color(self) -> str:
        return get_color(self.level)


@dataclass
class RiskRegister:
    organization: str
    assessor: str
    assessment_date: str
    risks: List[Risk] = field(default_factory=list)

    @classmethod
    def from_dict(cls, data: dict) -> "RiskRegister":
        register = cls(
            organization=data.get("organization", "Organization"),
            assessor=data.get("assessor", "Security Analyst"),
            assessment_date=data.get("assessment_date", datetime.utcnow().date().isoformat()),
        )
        for r in data.get("risks", []):
            register.risks.append(Risk(
                id=r.get("id", ""),
                title=r["title"],
                category=r.get("category", ""),
                likelihood=int(r["likelihood"]),
                impact=int(r["impact"]),
                description=r.get("description", ""),
                existing_controls=r.get("existing_controls", ""),
                mitigation=r.get("mitigation", ""),
                owner=r.get("owner", ""),
                due_date=r.get("due_date", ""),
                status=r.get("status", "open"),
            ))
        return register

    def sorted_risks(self) -> List[Risk]:
        return sorted(self.risks, key=lambda r: r.score, reverse=True)

    def summary(self) -> dict:
        levels = {"critical": 0, "high": 0, "medium": 0, "low": 0}
        for r in self.risks:
            levels[r.level] += 1
        return levels
